### PR TITLE
State revamp (So that we can see the diff)

### PR DESCRIPTION
**Pull request by pipermerriam**
Related Reddit Post: https://www.reddit.com/r/ethereum/comments/4r1k19/a_pyethereum_revampinprogress_purification/

This allows us to see the diff.

### CODE REVIEW

**Comment by veox** [comment]
Was going to write "anyone could have clicked Pull Request", but then remembered not everyone's logged in on github, and Travis would've been overworked.

File: ethereum/chain.py [hunk 1]
```diff
@@ -4,416 +4,401 @@
...
+
+    # Get the score (AKA total difficulty in PoW) of a given block
+    def get_score(self, block):
+        if not block:
+            return 0
+        key = 'score:'+block.header.hash
+        if key not in self.db:
+            try:
+                parent_score = self.get_score(self.get_parent(block))
+                self.db.put(key, str(parent_score + block.difficulty + random.randrange(10)))
```
Comment (janx) [inline ethereum/chain.py]: What's the point of `random.randrange(10)` here? @vbuterin

File: ethereum/chain.py [hunk 2]
```diff
@@ -4,416 +4,401 @@
...
-        if self.head_candidate:
-            return self.head_candidate.transaction_count
+            return None
+
+    # This should be called when a miner sees a transaction, and may
+    # potentially be interested in including it in a block
+    def add_transaction(self, tx):
+        if tx.gasprice >= self.min_gasprice:
+            i = 0
+            while i < len(self.transaction_queue) and tx.gasprice < self.transaction_queue[i]:
```
Comment (janx) [inline ethereum/chain.py]: Missing `.gasprice`? `tx.gasprice < self.transaction_queue[i].gasprice`

File: ethereum/casper_utils.py [hunk 3]
```diff
@@ -32,22 +32,38 @@ def get_contract_code(init_code):
...
 
 def get_finalizer_code():
-    import serpent
-    return get_contract_code(serpent.compile(open(finalizer_path).read()))
-
+    global _finalizer_code
+    if not _finalizer_code:
+        import serpent
+        _finalizer_code = get_contract_code(serpent.compile(open(finalizer_path).read()))
+    return _finalizer_code
```
Comment (janx) [inline ethereum/casper_utils.py]: :+1:

File: ethereum/parse_genesis_declaration.py [hunk 4]
```diff
@@ -43,8 +43,7 @@ def state_from_genesis_declaration(genesis_data, env, block=None):
                 state.set_storage_data(addr, parse_as_bin(k), parse_as_bin(v))
     initialize(state, block)
     state.commit()
-    # genesis block's state_root should be blank node hash
-    # Don't do this: block.header.state_root = state.trie.root_hash
+    block.header.state_root = state.trie.root_hash
```
Comment (janx) [inline ethereum/parse_genesis_declaration.py]: I think this change will break https://github.com/ethereum/pyethapp/pull/169.

**Comment by vbuterin** [comment]
> I think this change will break ethereum/pyethapp#169
<https://github.com/ethereum/pyethapp/pull/169>.

Why? It's a one line of code change that only happens when parsing genesis
block declarations.

> Missing .gasprice? tx.gasprice < self.transaction_queue[i].gasprice

Yep! Will fix.

> What's the point of random.randrange(10) here? @vbuterin
<https://github.com/vbuterin>

It's the selfish mining fix: randomly perturb the score of each block so
that in the event that you get blocks that otherwise have the same score,
you pick a random one.

**Pull request merged**